src/rust/ide/view/graph-editor/src/builtin/visualization/java_script/scatterPlot.js

```diff
@@ -43,111 +71,190 @@ class ScatterPlot extends Visualization {
-                    return x(d[headerNames[0]]);
-                })
-                .attr("cy", function (d) {
-                    return y(d[headerNames[1]]);
-                })
-                .attr("r", 8)
-                .style("fill", function (d) {
-                    return color(d[headerNames[2]])
-                })
-                .style("opacity", 0.5)
+                .append("text")
+                .text( d => d.label)
+                .attr('transform',d => "translate("+x(d.x)+","+y(d.y)+")")
+                .attr("font-size", "12px")
+                .attr("fill", "black");
+        }
 
+        ////////////////
+        /// Brushing ///
+        ////////////////
+
+        let brush = d3.brushX()
+            .extent([[0, 0], [width, height]])
+            .on("end", updateChart)
+
+
+        if (points.connected !== "yes") {
             scatter
                 .append("g")
                 .attr("class", "brush")
                 .call(brush);
+        }
 
-            var idleTimeout
+        let idleTimeout
 
-            function idled() {
-                idleTimeout = null;
-            }
+        function idled() {
+            idleTimeout = null;
+        }
 
-            function updateChart() {
-                let extent = d3.event.selection;
+        function updateChart() {
+            let extent = d3.event.selection;
 
-                if (!extent) {
-                    if (!idleTimeout) return idleTimeout = setTimeout(idled, 350);
-                    x.domain([4, 8])
-                } else {
-                    x.domain([x.invert(extent[0]), x.invert(extent[1])])
-                    scatter.select(".brush").call(brush.move, null)
-                }
+            if (!extent) {
+                if (!idleTimeout) return idleTimeout = setTimeout(idled, 350);
+                x.domain([xMin - dx, xMax + dx]);
+            } else {
+                x.domain([x.invert(extent[0]), x.invert(extent[1])]);
+                scatter.select(".brush").call(brush.move, null);
+            }
+
+            xAxis.transition().duration(1000).call(d3.axisBottom(x));
+            scatter
+                .selectAll("path")
+                .transition().duration(1000)
+                .attr('transform',d => "translate("+x(d.x)+","+y(d.y)+")")
 
-                xAxis.transition().duration(1000).call(d3.axisBottom(x))
+            if (points.labels === "visible") {
                 scatter
-                    .selectAll("circle")
+                    .selectAll("text")
                     .transition().duration(1000)
-                    .attr("cx", function (d) {
-                        return x(d[headerNames[0]]);
-                    })
-                    .attr("cy", function (d) {
-                        return y(d[headerNames[1]]);
-                    })
-
+                    .attr('transform',d => "translate("+x(d.x)+","+y(d.y)+")")
             }
-        };
+        }
```

Comment:
This is a really long function. Can you split it into functions instead of making headings?